### PR TITLE
Generalize Pill component

### LINKED ISSUE #3967: Text Sets: Generalize "Pill" filter UI
## Feature Description

Take the Pill filter UI used in the third party media pane and generalize it to be reused in the text pane.

- [ ] Generalize components to central location
- [ ] Fix deviations from design (Sam had some comments)
- [ ] Add unit tests
- [ ] Add to story book

https://www.figma.com/file/bMhG3KyrJF8vIAODgmbeqT/Stories-Stable?node-id=803%3A317

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

A text set panel with the title, pills, and arrow button that shows the top filter options or all of them depending on the toggle.

The top filter options are Title, Paragraph, Steps, List. The remaining options are Table, Quote, Contact.

You can only select one filter at a time and the bubble becomes outlined in blue.

## Implementation Brief

_Please fill this section before commencing issue resolution_.

### PR DESCRIPTION
**Pull request by miina**
## Summary
Creates `PillGroup` and `Pill` components for general use instead of specifically for Media only.

## To-do

- [x] Verify matching with design

## User-facing changes
N/A (Behind the feature flag)

## Testing Instructions
- Open 3rd party Media tab
- Verify that the Category Pills filter the Media as expected when clicking on it.
- Verify that it's possible to navigate in the categories using keyboard (up-down-left-right, selecting with Space/Enter)
- Verify that the selected Category is visibly different from the unselected ones.

---

Fixes #3967

### CODE REVIEW

File: assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js [hunk 1]
```diff
@@ -223,11 +222,11 @@ const Media3pCategories = ({
   );
 };
 
-Media3pCategories.propTypes = {
+PillGroup.propTypes = {
   categories: PropTypes.array.isRequired,
   selectedCategoryId: PropTypes.string,
   selectCategory: PropTypes.func,
   deselectCategory: PropTypes.func,
```
Comment (barklund) [inline assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js]: Perhaps `category` is not the best generic term here? Maybe just `item` or something similarly vague?

File: assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js [hunk 2]
```diff
@@ -84,7 +83,7 @@ const ExpandButton = styled(ArrowDown)`
   align-items: center;
 `;
 
-const Media3pCategories = ({
+const PillGroup = ({
```
Comment (barklund) [inline assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js]: I have quite a few things I'd like to see cleaned up in this one (didn't review it when it was originally added):

* This component has a few rather complex `useEffect`s - can we just move them to external hooks instead to reduce complexity?
* It has a static id for the pill group, which doesn't work if there's more than one pill group on the DOM at the same time.
* It has some `data-*` attributes too, that I'd prefer to see removed.
* It should have `Boolean(categories.length)` moved to a variable above the return and use a short-circuit conditional rather than ternary: `{hasCategories && <>...</>}` rather than `{hasCategories ? <>...</> : null}`
* I think the `role=tablist` requires a label - either by `aria-labelledby` if we have it in another element or just `aria-label` if we don't.
* The element between `tablist` and `tab` (the `CategoryPillInnerContainer`) should have a `role=presentation` to not confuse screen readers.
* I'm not sure about the purpose of the `renderCategories` function here. It is only used once anyway. It should either be made into a real component or just embedded inline in the JSX.
* It's a bit weird that `title` is sent as a parameter to the `<Pill />` component and not just as children and used as `<Pill ...>{e.displayName}</Pill>`.
* Is `displayName` the best name for the label of a pill? Maybe it should be renamed to just `name`, `title` or `label`.

File: assets/src/edit-story/components/library/panes/shared/pillGroup/pill.js [hunk 3]
```diff
@@ -13,49 +13,56 @@
...
     opacity: 0;
   }
 `;
 
-PillContainer.propTypes = {
-  isSelected: PropTypes.bool,
-};
-
-const CategoryPill = ({
+function Pill({
```
Comment (barklund) [inline assets/src/edit-story/components/library/panes/shared/pillGroup/pill.js]: A few comments about this component in general:

* I notice that it uses a few `data-*` attributes, that I think we can do without. 
* The key listeners don't belong in this scope. Both of them (roving tabindex and _down-to-expand_) should go in the parent component. Events will properly bubble up to the parent scope. Thus there's no need for the `setIsExpanded` or `isExpanded` props either on this one.
* What's with the `className`? Is it used for styling or?
* The only reason we're passing `index` here is to auto-select the first item if none is selected. We can do that by automatically selecting the first id in the parent `PillGroup` whenever the list content changes. And then remove `index` here completely - that would clean up both this and the parent component a bit.

**Review by barklund — COMMENTED** [review]
Now that we're touching these components, I'd like to see them generalized a bit more and some short-comings fixed.

File: assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js (hunk 2 above)
Comment (miina) [reply assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js]: Alright, sounds good, will review all that was already there as well.

File: assets/src/edit-story/components/library/panes/shared/pillGroup/pill.js (hunk 3 above)
Comment (miina) [reply assets/src/edit-story/components/library/panes/shared/pillGroup/pill.js]: > We can do that by automatically selecting the first id in the parent PillGroup whenever the list content changes

Can you elaborate on what you meant by "selecting the first id"? If you meant setting `isSelected` to `true` in case of the first ID then this would mean styling of the pill changing which would be incorrect.

File: assets/src/edit-story/components/library/panes/media/media3p/karma/mediaFetching.karma.js [hunk 4]
```diff
@@ -575,7 +571,7 @@ describe('Media3pPane fetching', () => {
 
     it('should handle pressing up', async () => {
       mockListMedia();
-      await fixture.events.click(media3pTab);
+      await fixture.events.click(fixture.editor.library.media3pTab);
```
Comment (barklund) [inline assets/src/edit-story/components/library/panes/media/media3p/karma/mediaFetching.karma.js:574]: It seems like the same 3 lines are repeating in all (or many of) these tests - can we move it to a `beforeEach`? Maybe group in a `define()` for the relevant ones.

Comment (miina) [reply assets/src/edit-story/components/library/panes/media/media3p/karma/mediaFetching.karma.js:574]: It feels like generally the file could be reworked to use `fixture.editor.library` more and to get rid of using the `data-testid` values, however, I'd prefer doing this in a separate PR since it's a bit unrelated to the Pill, perhaps we can create a code quality follow-up issue. WDYT?

File: assets/src/edit-story/components/library/panes/shared/pillGroup/useHandleRowVisibility.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    const selectedItemOffsetTop = selectedItem?.offsetTop || 0;
+
+    for (let pill of itemRefs.current) {
+      const isSameRow =
+        selectedItem && pill.offsetTop === selectedItemOffsetTop;
+      if (selectedItem && !isSameRow && !isExpanded) {
+        pill.classList.add('invisible');
+      } else {
+        pill.classList.remove('invisible');
+      }
```
Comment (barklund) [inline assets/src/edit-story/components/library/panes/shared/pillGroup/useHandleRowVisibility.js:70]: Classes? Could we convert this to S-C's vernacular easily or is that too much work? It would require setting an attribute in another array somewhere and using that in the loop of rendering items. It's only a nit, no worries if too tricky.

File: assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    selectedItemId,
+    setFocusedRowOffset,
+    itemRefs,
+  });
+
+  const hasItems = items.length > 0;
+  useRovingTabIndex({ ref: sectionRef }, [isExpanded]);
+  useKeyDownEffect(
+    sectionRef,
+    !isExpanded ? 'down' : [],
```
Comment (barklund) [inline assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js]: Actually this should be the empty string if no listener is required. I [added that recently](https://github.com/google/web-stories-wp/blob/267a960ee8d35e07fed32062dac546ef823d6ca1/assets/src/edit-story/components/keyboard/index.js#L96-L98) for this exact purpose.

File: assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    () => hasItems && setIsExpanded(true),
+    [isExpanded, hasItems]
+  );
```
Comment (barklund) [inline assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js]: One tiny nit: You can use the left arrow to move to an invisible option while the list is collapsed (if you have selected an item further down that the first item). Maybe this one should just expand on any arrow press and not only on down arrow?

**Review by barklund — APPROVED** [review]
I hadn't actually reviewed the experience of using keyboard for the categories - it works very nicely. Great job by the media pod there and even better generalization here.

Also, that `useRovingTabIndex` hook is very close to the thing [I was requesting elsewhere](https://github.com/google/web-stories-wp/pull/4462#discussion_r488205701) - seems promising to generalize even further in the future.

Other than that, this looks good. Let's get it in!

File: assets/src/edit-story/components/library/panes/media/media3p/karma/mediaFetching.karma.js (hunk 4 above)
Comment (barklund) [reply assets/src/edit-story/components/library/panes/media/media3p/karma/mediaFetching.karma.js:574]: Works for me.

File: assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js (hunk 7 above)
Comment (miina) [reply assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js]: Hmm, I'm not sure -- in this case, ideally, it should do the moving + expand at the same time, otherwise, if the user is on the first item and clicks `right`, it would just expand but not move. Just tested it out and it feels a bit unexpected.

I'll merge this as-is and we can consider/test it again when implementing the text sets panel.

Comment (barklund) [reply assets/src/edit-story/components/library/panes/shared/pillGroup/pillGroup.js]: That's fine - it's a tiny edge case.

File: assets/src/edit-story/components/library/panes/media/media3p/karma/mediaFetching.karma.js (hunk 4 above)
Comment (miina) [reply assets/src/edit-story/components/library/panes/media/media3p/karma/mediaFetching.karma.js:574]: Follow-up issue: https://github.com/google/web-stories-wp/issues/4536
